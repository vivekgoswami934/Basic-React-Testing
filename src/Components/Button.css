.button {
  margin-top: 10px;
  color: black;
  padding: 10px;
  cursor: pointer;
  border-radius: 5px;
}

.red {
  background-color: red;
}
.yellow {
  background-color: red;
  color: black;
}
.green {
  background-color: green;
  color: black;
}
.pink {
  background-color: pink;
  color: black;
}

.ghost {
  border: none;
}

.outline {
  border: 5px solid black;
}
