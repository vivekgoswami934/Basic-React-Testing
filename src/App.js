import { useState } from "react"
import './App.css';
import Button from "./Components/Button"

function App() {

  const [theme, setTheme] = useState("light")


  return (
    <div className="App">

      <h2>Theme is {theme}</h2>
      <div>Click Me</div>

      <Button colorScheme={"red"} variant={"ghost"} onClick={() => setTheme(theme==="light"?"dark":"light")} >Click Me</Button>

    </div>
  );
}

export default App;
