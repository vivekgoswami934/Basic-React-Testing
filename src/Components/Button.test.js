

import { fireEvent, render, screen } from "@testing-library/react"
import Button from "./Button"
import App from "../App";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer"


describe("Testing the custom Button", () => {

    it("should be present in the DOM", () => {

        render(<Button>Click Me</Button>);   // virtually
        //  screen.debug(); 
        let button = screen.getByText("Click Me");
        expect(button).toBeInTheDocument();

        // render(<App></App>)
        // let button = screen.getByText("Click Me");
        // expect(button).toBeInTheDocument();

    });


    it("Should have a button with text Click Me", () => {
        render(<App></App>)
        // screen.debug()
        let button = screen.getByTestId("customButton");  // single / a particular thing
        expect(button).toBeInTheDocument();
    })

    // Empty DOM element  <div></div>

    it("Should render without any children", () => {
        render(<Button></Button>);

        let button = screen.getByTestId("customButton")
        expect(button).toBeEmptyDOMElement();
    })


    it("Should toggle the theme", () => {
        render(<App></App>);
        let element = screen.getByText("Theme is light");

        expect(element).toHaveTextContent("light");

        let button = screen.getByTestId("customButton");
        fireEvent.click(button);
        // userEvent.click(button);
        expect(element).toHaveTextContent("dark");

    });


    // my function is trigger or not ?



    it("should call the given function", () => {

        const mockFunction = jest.fn();  // () => {} anonymous function ;

        render(<Button onClick={mockFunction}>Click Me</Button>);

        // screen.debug()

        const button = screen.getByTestId("customButton");

        userEvent.click(button);
        userEvent.click(button);

        // expect(mockFunction).toBeCalled();
        expect(mockFunction).toBeCalledTimes(2);

    })


    // snapshots

    it("Should match the snapshot", () => {

        const tree = renderer.create(
            <Button colorScheme={"green"} variant={"outlined"}>
                Increment Counter
            </Button>
        ).toJSON()  //           .toTree();
 
        expect(tree).toMatchSnapshot()

    })

});



// npm i react-test-renderer  ? ==> 

                      // why this snapshot helpful 
                          //   --> if someone change in structure in button 
                              //    ==> when the testing the button is going to  failed.

                // press u ==>update the failing the snapshot