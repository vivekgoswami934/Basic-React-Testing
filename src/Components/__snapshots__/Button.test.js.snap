// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the custom Button Should match the snapshot 1`] = `
<button
  className="button green outlined"
  data-testid="customButton"
>
  Increment Counter
</button>
`;
